### PR TITLE
[complete-doc]: add support for constuctor doc

### LINKED ISSUE #4781: completed-docs does not ask for constructor doc
### Bug Report

- __TSLint version__:
5.18.0
- __TypeScript version__:
3.2.4
- __Running TSLint via__: (pick one) CLI / Node.js API / VSCode / grunt-tslint / Atom / Visual Studio / etc
CLI

`completed-docs` does not ask for constructor documentation. Is it a bug?

### PR DESCRIPTION
**Pull request by tanmoyopenroot**
#### PR checklist

- [x] Addresses an existing issue: fixes #4781 
- [x] enhancement
- [x] Includes tests
- [ ] Documentation update

#### Overview of change:
Added new doctype => ARGUMENT_CONSTRUCTORS = "constructors"

Below is the descriptor

```javascript
    public static ARGUMENT_DESCRIPTOR_CONSTRUCTOR = {
        properties: {
            [DESCRIPTOR_TAGS]: {
                properties: {
                    [TAGS_FOR_CONTENT]: {
                        items: {
                            type: "string",
                        },
                        type: "object",
                    },
                    [TAGS_FOR_EXISTENCE]: {
                        items: {
                            type: "string",
                        },
                        type: "array",
                    },
                },
            },
            [DESCRIPTOR_PRIVACIES]: {
                enum: [ALL, PRIVACY_PRIVATE, PRIVACY_PROTECTED, PRIVACY_PUBLIC],
                type: "string",
            },
            [DESCRIPTOR_OVERLOADS]: {
                type: "boolean",
            },
        },
        type: "object",
    };
```

### CODE REVIEW

File: src/rules/completedDocsRule.ts [hunk 1]
```diff
@@ -499,6 +539,14 @@ function walk(context: Lint.WalkContext<ExclusionsMap>) {
             );
         }
 
+        if (tsutils.isConstructorDeclaration(node) && tsutils.isClassDeclaration(node.parent)) {
```
Comment (JoshuaKGoldberg) [inline src/rules/completedDocsRule.ts]:
```suggestion
        if (tsutils.isConstructorDeclaration(node)) {
```

There's no case I can think of where we'd care what the parent of a constructor node is. Class declarations and class expressions are both valid class types.

File: test/rules/completed-docs/constructors/overloads/default/test.ts.lint [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class Foo {
```
Comment (JoshuaKGoldberg) [inline test/rules/completed-docs/constructors/overloads/default/test.ts.lint]:
```suggestion
class Foo1 {
```

The TypeScript APIs might display some odd behavior around type inference with multiple classes of the same name. Since this file is checking edge cases of overloading, let's give each of these classes a unique name.

File: test/rules/completed-docs/constructors/overloads/true/test.ts.lint [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+class Foo {
+    protected constructor(i: number);
+    protected constructor(o: any) {}
+}
+
+[default]: Documentation must exist for constructors.
+[public]: Documentation must exist for public constructors.
+[private]: Documentation must exist for private constructors.
```
Comment (JoshuaKGoldberg) [inline test/rules/completed-docs/constructors/overloads/true/test.ts.lint]: Missing test case: none of the tests exercise the `protected` constructor case.

File: src/rules/completed-docs/constructorExclusion.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+export class ConstructorExclusion extends Exclusion<IConstructorExclusionDescriptor> {
+    public readonly privacies: Set<Privacy> = this.createSet(this.descriptor.privacies);
+
+    public excludes(node: ts.Node) {
+        return !this.shouldPrivacyBeDocumented(node);
+    }
+
+    private shouldPrivacyBeDocumented(node: ts.Node) {
```
Comment (JoshuaKGoldberg) [inline src/rules/completed-docs/constructorExclusion.ts]: Just nitpicking here: what are your thoughts on making this all one big `excludes` function? I don't see a lot of benefit to splitting this one into two. It's a small extra conceptual step + a few more lines of code.

The other exclusion classes have a bit more complex logic, so they had to.

File: test/rules/completed-docs/constructors/overloads/default/test.ts.lint (hunk 2 above)
Comment (JoshuaKGoldberg) [reply test/rules/completed-docs/constructors/overloads/default/test.ts.lint]: _(on second though, a better name might be `FooPublic` or something that describes what the class is testing...)_

Comment (tanmoyopenroot) [reply test/rules/completed-docs/constructors/overloads/default/test.ts.lint]: Done

File: test/rules/completed-docs/constructors/overloads/true/test.ts.lint (hunk 3 above)
Comment (tanmoyopenroot) [reply test/rules/completed-docs/constructors/overloads/true/test.ts.lint]: Done

File: src/rules/completedDocsRule.ts (hunk 1 above)
Comment (tanmoyopenroot) [reply src/rules/completedDocsRule.ts]: Done

File: src/rules/completed-docs/constructorExclusion.ts (hunk 4 above)
Comment (tanmoyopenroot) [reply src/rules/completed-docs/constructorExclusion.ts]: Done

**Comment by tanmoyopenroot** [comment]
> Code changes generally LGTM, thanks for sticking to the existing styles! A few small comments here and there; the biggest thing is that `protected` is missing testing.

Done
